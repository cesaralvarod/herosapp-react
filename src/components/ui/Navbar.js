import React, { useContext } from "react";
import { Link, NavLink, useHistory } from "react-router-dom";
import { AuthContext } from "../../auth/AuthContext";
import { types } from "../../types/types";

const Navbar = () => {
  const history = useHistory();

  const {
    user: { name },
    dispatch,
  } = useContext(AuthContext);

  const handleLogout = (event) => {
    event.preventDefault();
    dispatch({ type: types.logout });
    history.replace("/login");
  };

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <button
        className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarToggler"
        aria-controls="navbarToggler"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span className="navbar-toggler-icon"></span>
      </button>

      <Link className="navbar-brand" to="/">
        Heros App
      </Link>

      <div className="collapse navbar-collapse" id="navbarToggler">
        <ul className="navbar-nav mr-auto mt-2 mt-lg-0">
          <li className="nav-item active">
            <NavLink
              activeClassName="active"
              className="nav-item nav-link"
              exact
              to="/marvel"
            >
              Marvel
            </NavLink>
          </li>
          <li className="nav-item">
            <NavLink
              activeClassName="active"
              className="nav-item nav-link"
              exact
              to="/dc"
            >
              DC
            </NavLink>
          </li>
          <li className="nav-item">
            <NavLink
              activeClassName="active"
              className="nav-item nav-link"
              exact
              to="/search"
            >
              Search
            </NavLink>
          </li>
        </ul>
        <ul className="navbar-nav mt-2" style={{ marginLeft: "auto" }}>
          <li className="nav-item">
            <span className="text-info nav-item nav-link">{name}</span>
          </li>
          <li className="nav-item">
            <button className="btn nav-item nav-link" onClick={handleLogout}>
              Logout
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
