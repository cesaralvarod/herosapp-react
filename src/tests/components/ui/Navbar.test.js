import React from "react";
import { mount } from "enzyme";
import { AuthContext } from "../../../auth/AuthContext";
import Navbar from "../../../components/ui/Navbar";
import { MemoryRouter, Router } from "react-router-dom";
import { types } from "../../../types/types";

describe("Testing in Navbar component", () => {
  const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
  };

  const contextValue = {
    user: { name: "Cesar", logged: true },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <MemoryRouter>
      <AuthContext.Provider value={contextValue}>
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </AuthContext.Provider>
    </MemoryRouter>
  );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("Component should show correctly", () => {
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("nav.navbar span.nav-link").text().trim()).toBe(
      contextValue.user.name
    );
  });

  test("Component should simulate logout and call history", () => {
    const btnLogout = wrapper.find("button.btn");

    btnLogout.simulate("click", { preventDefault: jest.fn() });

    expect(contextValue.dispatch).toHaveBeenCalledWith({ type: types.logout });

    expect(historyMock.replace).toHaveBeenCalledWith("/login");
  });
});
