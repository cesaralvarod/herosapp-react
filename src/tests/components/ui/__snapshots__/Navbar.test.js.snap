// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in Navbar component Component should show correctly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <button
    aria-controls="navbarToggler"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarToggler"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Heros App
  </a>
  <div
    className="collapse navbar-collapse"
    id="navbarToggler"
  >
    <ul
      className="navbar-nav mr-auto mt-2 mt-lg-0"
    >
      <li
        className="nav-item active"
      >
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          Marvel
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          DC
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          aria-current={null}
          className="nav-item nav-link"
          onClick={[Function]}
        >
          Search
        </a>
      </li>
    </ul>
    <ul
      className="navbar-nav mt-2"
      style={
        Object {
          "marginLeft": "auto",
        }
      }
    >
      <li
        className="nav-item"
      >
        <span
          className="text-info nav-item nav-link"
        >
          Cesar
        </span>
      </li>
      <li
        className="nav-item"
      >
        <button
          className="btn nav-item nav-link"
          onClick={[Function]}
        >
          Logout
        </button>
      </li>
    </ul>
  </div>
</nav>
`;
